File: client-python/examples/advanced_usage.py
#!/usr/bin/env python3
"""
Advanced usage example for slskr Python client
- Error handling
- Retries and timeouts
- Context managers
- Concurrent operations
"""

import asyncio
from slskr import SlskrClient, ApiError, NetworkError, TimeoutError


async def search_with_retry(client, query, max_retries=3):
    """Search with automatic retry on failure"""
    
    for attempt in range(max_retries):
        try:
            result = await client.create_search(query=query)
            print(f"✓ Search created: {result.get('id')}")
            return result
        except ApiError as e:
            if e.is_not_found():
                print(f"Search endpoint not found")
                return None
            elif e.is_unauthorized():
                print("Invalid API key")
                return None
            print(f"Attempt {attempt + 1} failed: {e}")
            if attempt < max_retries - 1:
                await asyncio.sleep(2 ** attempt)  # Exponential backoff
        except TimeoutError:
            print(f"Timeout on attempt {attempt + 1}")
            if attempt < max_retries - 1:
                await asyncio.sleep(1)
        except NetworkError as e:
            print(f"Network error: {e}")
            if attempt < max_retries - 1:
                await asyncio.sleep(2)
    
    return None


async def concurrent_searches(client, queries):
    """Execute multiple searches concurrently"""
    
    tasks = [
        search_with_retry(client, q)
        for q in queries
    ]
    
    results = await asyncio.gather(*tasks, return_exceptions=True)
    
    successful = [r for r in results if r is not None and not isinstance(r, Exception)]
    print(f"\nCompleted {len(successful)}/{len(queries)} searches")
    
    return successful


async def fetch_all_results(client, search_id, batch_size=50):
    """Fetch all search results with pagination"""
    
    all_results = []
    offset = 0
    
    while True:
        details = await client.get_search_details(search_id, limit=batch_size)
        results = details.get("results", [])
        
        if not results:
            break
        
        all_results.extend(results)
        offset += len(results)
        print(f"Fetched {offset} results...")
        
        # Stop if we got less than batch_size (indicates end of results)
        if len(results) < batch_size:
            break
    
    return all_results


async def main():
    """Advanced usage examples"""
    
    # Using context manager
    async with SlskrClient(
        base_url="http://127.0.0.1:5030",
        token="your-api-key-here",
        timeout=30,
        retries=3,
        debug=True
    ) as client:
        
        try:
            # Example 1: Single search with retry
            print("=== Example 1: Search with Retry ===")
            search = await search_with_retry(client, "debussy prelude")
            
            if search:
                search_id = search.get('id')
                
                # Example 2: Fetch all results with pagination
                print("\n=== Example 2: Paginated Results ===")
                all_results = await fetch_all_results(client, search_id, batch_size=50)
                print(f"Total results: {len(all_results)}")
                
                # Example 3: Process results
                if all_results:
                    print("\n=== Example 3: Process Results ===")
                    for result in all_results[:5]:
                        print(f"  {result.get('filename')} ({result.get('filesize')} bytes)")
            
            # Example 4: Concurrent operations
            print("\n=== Example 4: Concurrent Searches ===")
            queries = ["radiohead", "pink floyd", "david bowie", "the beatles"]
            searches = await concurrent_searches(client, queries)
            
            # Example 5: Batch operations
            print("\n=== Example 5: Batch Operations ===")
            batch = client.batch.builder()
            batch.get("/api/stats", op_id="stats")
            batch.get("/api/config", op_id="config")
            batch.get("/api/capabilities", op_id="caps")
            
            response = await batch.execute()
            print(f"Batch results: {len(response.get_successful())} successful")
            
            # Example 6: Error handling in batch
            print("\n=== Example 6: Error Handling ===")
            batch2 = client.batch.builder()
            batch2.get("/api/invalid-endpoint", op_id="invalid")
            batch2.get("/api/stats", op_id="valid")
            
            response2 = await batch2.execute()
            for result in response2.results:
                if result.is_error():
                    print(f"Operation {result.id} failed with status {result.status}")
                else:
                    print(f"Operation {result.id} succeeded")
            
        except Exception as e:
            print(f"Unexpected error: {e}")
            import traceback
            traceback.print_exc()


if __name__ == "__main__":
    asyncio.run(main())

File: client-go/examples/advanced_usage.go
//go:build examples

package main

import (
	"context"
	"fmt"
	"log"
	"sync"
	"time"

	"github.com/snapetech/slskr/client-go"
)

func searchWithContext(ctx context.Context, client *slskr.Client, query string) (map[string]interface{}, error) {
	// Create a timeout context if not already set
	if _, ok := ctx.Deadline(); !ok {
		var cancel context.CancelFunc
		ctx, cancel = context.WithTimeout(ctx, 30*time.Second)
		defer cancel()
	}

	return client.CreateSearch(ctx, query)
}

func concurrentSearches(ctx context.Context, client *slskr.Client, queries []string) {
	var wg sync.WaitGroup
	results := make(chan map[string]interface{}, len(queries))

	for _, q := range queries {
		wg.Add(1)
		go func(query string) {
			defer wg.Done()
			result, err := searchWithContext(ctx, client, query)
			if err != nil {
				fmt.Printf("Error searching '%s': %v\n", query, err)
				return
			}
			results <- result
		}(q)
	}

	go func() {
		wg.Wait()
		close(results)
	}()

	count := 0
	for result := range results {
		count++
		fmt.Printf("Search created: %v\n", result["id"])
	}
	fmt.Printf("Completed %d/%d searches\n", count, len(queries))
}

func listAllResults(ctx context.Context, client *slskr.Client, searchID string) {
	limit := 50
	offset := 0
	total := 0

	for {
		result, err := client.ListSearches(ctx, limit, offset)
		if err != nil {
			fmt.Printf("Error fetching results: %v\n", err)
			break
		}

		if len(result) == 0 {
			break
		}

		total += len(result)
		offset += len(result)

		fmt.Printf("Fetched %d results (total: %d)...\n", len(result), total)

		if len(result) < limit {
			break
		}
	}

	fmt.Printf("Total results: %d\n", total)
}

func batchOperationsExample(ctx context.Context, client *slskr.Client) {
	batch := client.NewBatchBuilder()

	// Add various operations
	batch.Get("/api/stats", nil)
	batch.Get("/api/config", nil)
	batch.Get("/api/capabilities", nil)

	batch.Post("/api/searches", map[string]interface{}{
		"query": "radiohead ok computer",
	}, nil)

	batch.Post("/api/searches", map[string]interface{}{
		"query": "pink floyd dark side",
	}, nil)

	fmt.Printf("Executing batch of %d operations...\n", batch.Size())

	response, err := batch.Execute(ctx)
	if err != nil {
		log.Printf("Batch error: %v\n", err)
		return
	}

	fmt.Printf("Batch completed in %dms\n", response.TotalTimeMs)
	fmt.Printf("Successful: %d\n", len(response.GetSuccessful()))
	fmt.Printf("Failed: %d\n", len(response.GetFailed()))
}

func errorHandlingExample(ctx context.Context, client *slskr.Client) {
	fmt.Println("\n=== Error Handling Example ===")

	// Try invalid endpoint in batch
	batch := client.NewBatchBuilder()
	batch.Get("/api/invalid-endpoint", nil)
	batch.Get("/api/stats", nil)

	response, err := batch.Execute(ctx)
	if err != nil {
		fmt.Printf("Batch execution failed: %v\n", err)
		return
	}

	for _, result := range response.Results {
		if result.IsError() {
			fmt.Printf("Operation %s failed with status %d\n", result.ID, result.Status)
		} else {
			fmt.Printf("Operation %s succeeded (status %d)\n", result.ID, result.Status)
		}
	}
}

func main() {
	client := slskr.NewClient("http://127.0.0.1:5030", "your-api-key-here")
	client.HTTPClient.Timeout = 30 * time.Second

	ctx := context.Background()

	fmt.Println("=== Advanced Usage Examples ===\n")

	// Example 1: Concurrent searches
	fmt.Println("1. Concurrent Searches")
	queries := []string{"beethoven", "mozart", "bach", "brahms"}
	concurrentSearches(ctx, client, queries)

	// Example 2: Batch operations
	fmt.Println("\n2. Batch Operations")
	batchOperationsExample(ctx, client)

	// Example 3: Error handling
	errorHandlingExample(ctx, client)

	// Example 4: List all with pagination
	fmt.Println("\n4. Pagination Example")
	listAllResults(ctx, client, "")

	// Example 5: Room operations
	fmt.Println("\n5. Room Operations")
	rooms, err := client.ListRooms(ctx)
	if err != nil {
		fmt.Printf("Error listing rooms: %v\n", err)
	} else {
		fmt.Printf("Available rooms: %d\n", len(rooms))
	}

	// Example 6: Shares
	fmt.Println("\n6. Share Management")
	shares, err := client.ListShares(ctx, 10, 0)
	if err != nil {
		fmt.Printf("Error listing shares: %v\n", err)
	} else {
		fmt.Printf("Shared files: %d\n", len(shares))
	}

	// Example 7: Filters
	fmt.Println("\n7. Search Filters")
	filters, err := client.GetFilters(ctx)
	if err != nil {
		fmt.Printf("Error getting filters: %v\n", err)
	} else {
		fmt.Printf("Filters: %v\n", filters)
	}
}
